crates/schema/src/def/validate/v9.rs

```diff
@@ -1347,11 +1397,47 @@ mod tests {
             .finish();
         let result: Result<ModuleDef> = builder.finish().try_into();
 
-        expect_error_matching!(result, ValidationError::OnlyBtree { index } => {
+        expect_error_matching!(result, ValidationError::HashIndexUnsupported { index } => {
             &index[..] == "Bananas_b_idx_hash"
         });
     }
 
+    #[test]
+    fn unique_constrain_without_index() {
+        let mut builder = RawModuleDefV9Builder::new();
+        builder
+            .build_table_with_new_type(
+                "Bananas",
+                ProductType::from([("b", AlgebraicType::U16), ("a", AlgebraicType::U64)]),
+                false,
+            )
+            .with_unique_constraint(1)
+            .finish();
+        let result: Result<ModuleDef> = builder.finish().try_into();
+
+        expect_error_matching!(result, ValidationError::UniqueConstraintWithoutIndex { constraint } => {
+            &**constraint == "Bananas_a_key"
+        });
+    }
+
+    #[test]
+    fn direct_index_only_u8_to_u64() {
+        let mut builder = RawModuleDefV9Builder::new();
+        builder
+            .build_table_with_new_type(
+                "Bananas",
+                ProductType::from([("b", AlgebraicType::I32), ("a", AlgebraicType::U64)]),
+                false,
+            )
+            .with_index(RawIndexAlgorithm::Direct { column: 0.into() }, "bananas_b")
+            .finish();
+        let result: Result<ModuleDef> = builder.finish().try_into();
+
+        expect_error_matching!(result, ValidationError::DirectIndexOnNonUnsignedInt { index, .. } => {
+            &index[..] == "Bananas_b_idx_direct"
+        });
```

Comment:
👍

Reply:
You'll also want to add a direct index to `valid_definition` & make sure it goes through